### PR TITLE
node: Flow cancel enhancements and bug fixes

### PR DESCRIPTION
**Pull request by johnsaigle**
It may be helpful to read commits individually while reviewing.

### Bug fixes

- Flow cancel did not trigger when pending transfers were released be3696a2b4354457e779b495e9e7a0b3269afa30
- Flow cancel did not trigger when loading pending transfers from the database 0524d8b6e3b61f9004a59cecb395c65ef9a0960a
- Changed error handling to prevent "crash looping" when Governor usage exceeds daily limit 013d79acb47ed1e561aa0ab72ce3ad8f2e2bc68a

### Additional features

- **Toggle Flow Cancel**: Added a command-line flag to enable or disable flow-canceling when starting the node. This should allow Guardians to disable flow canceling in the case of future bugs or during a security incident. This should prevent the need to rollback to earlier Guardian versions. (@mdulin2 ) 9120d17760c89d215ef43a0279d45c1ae3ae8825
- **Deterministic iteration**: Use deterministic ordering when iterating over chain entries and modifying their state. 38c79eef5f0f543dfbf3de33209472729bb8c042
- **Addional metrics**: Modify the monitoring code and protobuf files to make the status of the Governor more legible when flow-canceling is enabled. This can be consumed by Wormhole Dashboard to better reflect the effects of flow cancelling. a29158ce72b0ea19ee9e6844d65b3cab6f5cd2ff https://github.com/wormhole-foundation/wormhole/commit/eb13ff9db9029886bf8d138932d094ee9ee8054d

### CODE REVIEW

File: node/pkg/governor/governor_test.go
Comment (johnsaigle) [inline node/pkg/governor/governor_test.go:1795]: Note to self: delete this. It's from Jason's PR

File: node/pkg/governor/governor_test.go
Comment (johnsaigle) [reply node/pkg/governor/governor_test.go:1795]: Ah this came up during rebasing. I changed the branch target a few times. I'll restore this section

File: node/pkg/governor/governor_monitoring.go
Comment (johnsaigle) [reply node/pkg/governor/governor_monitoring.go:265]: I prefer the verbosity here. It took me a long time to sort through the different flow cancel edge-cases and I want it to be really clear going forward. Even if the variable name is long I think the concept it's describing represents a net reduction in mental effort for understanding what the code does

File: node/pkg/governor/governor_test.go
Comment (johnsaigle) [reply node/pkg/governor/governor_test.go]: Let me know if this looks better from your side now

**Review by pleasew8t — APPROVED** [review]

File: node/pkg/governor/governor.go
Comment (bruce-riley) [inline node/pkg/governor/governor.go:138]: What if we put out a release that reduces the big transfer size and there is something in the 24 hour window that exceeds the new BTS? Would that violate this constraint? I tried looking through the code, but I don't see where that's handled. I could've missed something though.

File: node/pkg/governor/governor.go
Comment (bruce-riley) [inline node/pkg/governor/governor.go:194]: If you make this a slice of `*chainEntry` you would not have to do map look ups everywhere you iterate through it. Of course they you should probably rename it to something like `chainsById`.

File: node/pkg/governor/governor.go
Comment (bruce-riley) [inline node/pkg/governor/governor.go:699]: Doesn't restart the whole node, just the processor (which is still very bad).

File: node/pkg/governor/governor.go
Comment (bruce-riley) [inline node/pkg/governor/governor.go:689]: Nit: "CheckPendingForTime checks if a pending message is ready to be released, and if so, modifies..."

File: node/pkg/governor/governor.go
Comment (bruce-riley) [inline node/pkg/governor/governor.go:717]: Since this happens on the processor routine, I think it would definitely be better to avoid these map look ups and replace `gov.chainIds` with a slice of pointers.

File: node/pkg/governor/governor.go
Comment (bruce-riley) [inline node/pkg/governor/governor.go:829]: Nit: "processor" not "process".

Same comment below.

File: node/pkg/governor/governor.go
Comment (bruce-riley) [inline node/pkg/governor/governor.go:932]: Seems like a program bug, so maybe panic is better?

File: node/pkg/governor/governor_test.go
Comment (bruce-riley) [inline node/pkg/governor/governor_test.go:189]: I think it should panic because this would be a program bug, right?

File: node/pkg/governor/governor_db.go
Comment (johnsaigle) [reply node/pkg/governor/governor_db.go:255]: Sure I'm not opposed to upgrading the severity to Error.

File: node/pkg/governor/governor_prices.go
Comment (johnsaigle) [reply node/pkg/governor/governor_prices.go:312]: Yes, thanks!

File: proto/gossip/v1/gossip.proto
Comment (johnsaigle) [reply proto/gossip/v1/gossip.proto:159]: Sure I'll take a look at this and get back to you

File: node/pkg/governor/governor.go
Comment (johnsaigle) [reply node/pkg/governor/governor.go:932]: Do you know the effects this would have on the processor? I want to avoid a situation where a bug in the Governor logic causes it or related processes to restart. I see your point that this is a serious problem we don't want to encounter but I'm hesitant to add panics or otherwise affect liveness

File: node/pkg/governor/governor.go
Comment (johnsaigle) [reply node/pkg/governor/governor.go:194]: Do you mean changing `chains` from a map to a slice and removing `chainIds`?

File: node/pkg/governor/governor.go
Comment (johnsaigle) [reply node/pkg/governor/governor.go:699]: Thanks I'll update the comment!

File: node/pkg/governor/governor.go
Comment (bruce-riley) [reply node/pkg/governor/governor.go:138]: I guess I wasn't as concerned about "governor wonkiness" like the limits being a bit messed up until it clears. I just wanted to make sure we wouldn't panic or kill the processor in that case. If you feel that's covered, okay. (Or maybe have a test to verify that. :grin:).

File: node/pkg/governor/governor.go
Comment (bruce-riley) [reply node/pkg/governor/governor.go:194]: No, keep `chains` as is so we can do a map look up by chainID. Just replace `chainIds` with `chainsById` where the payload would be `*chainEntry` so you could use it to directly access the chain entry without doing that extra map lookup.

File: node/pkg/governor/governor.go
Comment (bruce-riley) [reply node/pkg/governor/governor.go:932]: Returning an error would cause the processor to restart, which is a silent failure unless someone is watching the logs.

A panic would kill the application, which should be much more visible.

My point was is that this should only happen in the case of a program bug, as in someone broke the code, which should be detected during testing, and never make it to production.

File: node/pkg/governor/governor.go
Comment (johnsaigle) [reply node/pkg/governor/governor.go:138]: Ah got it. The only remaining case where the processor should restart is here: https://github.com/wormhole-foundation/wormhole/pull/4016/files#diff-f68b867da294e995f0449fff8116515f60d0e718da580251c6fab60f5c4ce4abR277

This represents the sum of the transfers overflowing the bounds of int64

File: node/pkg/governor/governor.go
Comment (johnsaigle) [reply node/pkg/governor/governor.go:932]: The code that calls should `TrimAndSumValueForChain` log and swallows the error in all cases rather than propagating it up to the processor. (This is the intention anyway, let me know if you see otherwise.)

The idea here wasn't too fail silently or loudly in the sense of causing the processor to restart. Instead the error is logged.

My understanding is that the only path to restart the processor is via [returning an error from `CheckPending`](https://github.com/wormhole-foundation/wormhole/blob/80c26eb8729a4aab5ff338c1ac01a526531ead50/node/pkg/processor/processor.go#L264-L267). Errors from `ProcessMsg` do not cause a restart; [instead there's a continue.](https://github.com/wormhole-foundation/wormhole/blob/80c26eb8729a4aab5ff338c1ac01a526531ead50/node/pkg/processor/processor.go#L231)

File: node/pkg/governor/governor.go
Comment (johnsaigle) [reply node/pkg/governor/governor.go:194]: Could we break this out into a separate issue and not have it block the PR?
I see how it might be more optimal to iterate over pointers directly but I think having both data structures also means that we need to keep the pointers in sync.

I tried making this change and it broke a lot of things in the test suite. I think the reason is that there are helper functions there to update the `chainEntry` pointer in the map, which is just fine. But it means every time we do that we also need to update the `chainsById` slice. In the actual governor code we don't update the `chainEntry` pointer in the map but it feels like an accident waiting to happen. 

My preference would be to keep it simple and iterate over a static list of ~30 ints rather than worry about keeping these two things in sync.

File: proto/gossip/v1/gossip.proto
Comment (johnsaigle) [reply proto/gossip/v1/gossip.proto:159]: Added here: https://github.com/wormhole-foundation/wormhole/pull/4016/commits/eb13ff9db9029886bf8d138932d094ee9ee8054d

File: node/pkg/governor/governor.go
Comment (bruce-riley) [reply node/pkg/governor/governor.go:193]: I do feel that it would be better to eliminate a bunch of unnecessary map look ups, especially in any code called from the main processor routine, but I'm okay with deferring that to a future PR.

Comment (johnsaigle) [reply node/pkg/governor/governor.go:193]: Thanks I appreciate the flexibility. I'll write up a GitHub Issue summarizing the above

File: node/pkg/governor/governor.go
Comment (johnsaigle) [reply node/pkg/governor/governor.go:720]: I've added https://github.com/wormhole-foundation/wormhole/issues/4029 so we can discuss future performance enhancements

Comment (johnsaigle) [reply node/pkg/governor/governor.go:720]: #4029 recorded as an Issue here

File: node/pkg/governor/governor_prices.go
Comment (johnsaigle) [reply node/pkg/governor/governor_prices.go:312]: Just taking a look at this more closely. I don't think the check query test function interacts with flow cancelling at all. The flow cancel token list doesn't store price information, only the OriginChain, OriginAddress, and Symbol.

File: node/pkg/governor/governor_test.go
Comment (johnsaigle) [reply node/pkg/governor/governor_test.go:189]: Reposting [a related comment](https://github.com/wormhole-foundation/wormhole/pull/4016#discussion_r1683201546) for easier reviewing:

> The code that calls should TrimAndSumValueForChain log and swallows the error in all cases rather than propagating it up to the processor. (This is the intention anyway, let me know if you see otherwise.)
The idea here wasn't too fail silently or loudly in the sense of causing the processor to restart. Instead the error is logged.
My understanding is that the only path to restart the processor is via [returning an error from CheckPending](https://github.com/wormhole-foundation/wormhole/blob/80c26eb8729a4aab5ff338c1ac01a526531ead50/node/pkg/processor/processor.go#L264-L267). Errors from ProcessMsg do not cause a restart; [instead there's a continue.](https://github.com/wormhole-foundation/wormhole/blob/80c26eb8729a4aab5ff338c1ac01a526531ead50/node/pkg/processor/processor.go#L231)

File: node/pkg/governor/governor_prices.go
Comment (bruce-riley) [reply node/pkg/governor/governor_prices.go:312]: What I meant is that if the flow cancel token list was bad (it returns an error if `vaa.StringToAddress` fails), passing in true above would cause `check_query` to fail, allowing the problem to be caught earlier.

Comment (johnsaigle) [reply node/pkg/governor/governor_prices.go:312]: Ah I see. Sure, no harm adding it

**Review by pleasew8t — APPROVED** [review]

File: node/pkg/governor/governor.go
Comment (djb15) [inline node/pkg/governor/governor.go:840]: What's the purpose of the assignment here? Especially if the processor is going to die?

File: node/pkg/governor/governor_monitoring.go
Comment (djb15) [inline node/pkg/governor/governor_monitoring.go:326]: I don't think this is true if flow cancel was on but is then switched off, right? I think keeping the warnings for this is fine (since this toggling of flow cancel from on to off is an edge case), but probably worth clarifying in the comment?

**Review by bruce-riley — DISMISSED** [review]

File: node/pkg/governor/governor.go
Comment (johnsaigle) [reply node/pkg/governor/governor.go:840]: Fair point. It's not very clear from the diff but I think I just moved (old) lines from 765-767 down here. Agreed it seems unnesscary

**Review by djb15 — APPROVED** [review]

**Review by djb15 — DISMISSED** [review]

**Review by djb15 — APPROVED** [review]

**Review by bruce-riley — APPROVED** [review]

File: proto/gossip/v1/gossip.proto
Comment (evan-gray) [inline proto/gossip/v1/gossip.proto:186]: Just confirming, should this be here instead of under `Chain` alongside `remaining_available_notional`? Does each emitter have it's own set of notional values in a flow cancel world?

**Review by evan-gray — APPROVED** [review]
Reviewed proto. Approved pending confirmation of proto change.

File: proto/gossip/v1/gossip.proto
Comment (johnsaigle) [reply proto/gossip/v1/gossip.proto:186]: Yeah maybe under `Chain` makes more sense especially since that's where the existing `notional` measure is. I'll add a new commit.

Comment (evan-gray) [reply proto/gossip/v1/gossip.proto:186]: Okay, changing my review to request changes in the interim

**Review by evan-gray — CHANGES_REQUESTED** [review]
Holding for proto change.

**Review by evan-gray — APPROVED** [review]
reviewed proto

**Review by bruce-riley — APPROVED** [review]

**Pull request merged**
